### PR TITLE
fix: Fix evironment metadata N+1 for environments list

### LINKED ISSUE #2900: N+1 metadata query for `/api/v1/environments/?project=<id>`
Ref: https://flagsmith.sentry.io/issues/4502384230/?project=5544478&referrer=issue-stream&statsPeriod=90d&stream_index=4

### PR DESCRIPTION
**Pull request by zachaysan**
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

Isolated the cause of an N+1 query and implemented a fix without forcing it on every caller. We may want to consider reviewing the callers and seeing which of them load metadata as well, since the bulk of the performance enhancement is already complete.

## How did you test this code?

One new test. During the testing process I also manually inspecting the types of queries that were generated to ensure that they matched those reported by Sentry.

### CODE REVIEW

File: api/environments/tests/test_views.py [hunk 1]
```diff
@@ -536,6 +539,53 @@ def test_create_environment_without_required_metadata_returns_400(
...
+    environment_content_type: ContentType,
+) -> None:
+    # Given
+    with_environment_permissions([VIEW_ENVIRONMENT])
+
+    url = reverse("api-v1:environments:environment-list")
+    data = {"project": project.id}
+
+    # When
+    expected_query_count = 7
```
Comment (khvn26) [inline api/environments/tests/test_views.py]: nit: this line belongs to the `# Given` section

File: api/environments/tests/test_views.py [hunk 2]
```diff
@@ -536,6 +539,53 @@ def test_create_environment_without_required_metadata_returns_400(
     assert "Missing required metadata field" in response.json()["metadata"][0]
 
 
+def test_view_environment_with_staff_track_query_count(
```
Comment (khvn26) [inline api/environments/tests/test_views.py]: I'd prefer a slightly more structured test name here:

```suggestion
def test_view_environment_with_staff__query_count_is_expected(
```

File: api/environments/tests/test_views.py [hunk 3]
```diff
@@ -536,6 +539,53 @@ def test_create_environment_without_required_metadata_returns_400(
...
+    staff_client: FFAdminUser,
+    environment: Environment,
+    with_environment_permissions: Callable,
```
Comment (khvn26) [inline api/environments/tests/test_views.py]: nit: we should type this properly. Looks like it should be

```suggestion
    with_environment_permissions: Callable[[list[str], int], None],
```

I can see the fixture definition is also not typed correctly.

File: api/environments/tests/test_views.py [hunk 4]
```diff
@@ -536,6 +539,53 @@ def test_create_environment_without_required_metadata_returns_400(
...
+    project: Project,
+    django_assert_num_queries: Callable,
```
Comment (khvn26) [inline api/environments/tests/test_views.py]:
```suggestion
    django_assert_num_queries: Callable[[int], None],
```

File: api/permissions/permission_service.py [hunk 5]
```diff
@@ -130,7 +131,16 @@ def get_permitted_environments_for_user(
...
+    queryset = Environment.objects.filter(filter_)
+    if prefetch_metadata:
+        queryset = queryset.prefetch_related("metadata")
+
+    # Description is defered due to Oracle support where a
+    # query can't have a where clause if description is in
+    # the select parameters. This leads to an N+1 query for
+    # lists of environments when description is included, as
+    # each environment object re-queries the DB seperately.
+    return queryset.distinct().defer("description")
```
Comment (khvn26) [inline api/permissions/permission_service.py:143]: Maybe we could branch out the logic here based on `connection.vendor`?

Comment (zachaysan) [reply api/permissions/permission_service.py:143]: I was thinking that too, though it does open up a can of worms with respect to complexity.

File: api/environments/tests/test_views.py [hunk 6]
```diff
@@ -536,6 +539,53 @@ def test_create_environment_without_required_metadata_returns_400(
     assert "Missing required metadata field" in response.json()["metadata"][0]
 
 
+def test_view_environment_with_staff_track_query_count(
+    staff_client: FFAdminUser,
```
Comment (gagantrivedi) [inline api/environments/tests/test_views.py]: This should be `APIClient`

**Review by gagantrivedi — CHANGES_REQUESTED** [review]
Think we should try to add new tests under the `tests` directory only. The tests here should live under `/tests/unit/environment/test_views`, right @matthewelwell ?

File: api/permissions/permission_service.py (hunk 5 above)
Comment (matthewelwell) [reply api/permissions/permission_service.py:143]: Hmm... yeah, this is interesting. I guess for now, let's limit the scope of this PR to the N+1 caused by metadata. I think probably the better way to tackle this issue is to remove the `"description"` from wherever we list the environments. I've added this to the V2 API GH issue [here](https://github.com/Flagsmith/flagsmith/issues/1667) but we may be able to achieve this before then by creating a new endpoint perhaps.

**Comment by zachaysan** [comment]
> Think we should try to add new tests under the `tests` directory only. The tests here should live under `/tests/unit/environment/test_views`, right @matthewelwell ?

Sounds good. I've moved all the tests into the target file, including the one I wrote.

**Review by matthewelwell — APPROVED** [review]

**Review by gagantrivedi — APPROVED** [review]

**Pull request merged**